File: go/hamming/hamming.go
package hamming

import (
	"fmt"
)

// Distance calculates the hamming distance between two DNA strands.
func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("strings should be of same length")
	}
	ar, br := []rune(a), []rune(b)

	var distance int
	for i := 0; i < len(ar); i++ {
		if ar[i] != br[i] {
			distance++
		}
	}
	return distance, nil
}

File: c/hamming/src/hamming.c
#include "hamming.h"
#include <string.h>

long long compute(const char *lhs, const char *rhs)
{
    int length = 0;
    while(1)
    {
        if (lhs[length] == '\0' && rhs[length] == '\0')
        {
            break;
        }
        if (lhs[length] == '\0' || rhs[length] == '\0')
        {
            return -1;
        }
        length++;
    }

    long long distance = 0;
    for (int i = 0; i < length; i++)
    {
        if (lhs[i] != rhs[i])
        {
            distance++;
        }
    }
    return distance;
}
